src/lib/table/table-data-source.ts

```diff
@@ -0,0 +1,209 @@
+   */
+  sortingDataAccessor = (data: T, sortHeaderId: string): string|number => {
+    const value: number|string = data[sortHeaderId];
+    return isNaN(+value) ? value : +value;
+  }
+
+  /**
+   * Transforms data objects into a filter term that will be used to check against the filter if
+   * a filter is set. By default, the function will iterate over the values of the data object
+   * and convert them to a lowercase string.
+   * @param data Data object to convert to a string that checked for containing the filter term.
+   */
+  filterTermAccessor: ((data: T) => string) = (data: T): string => {
+    const accumulator = (currentTerm, key) => currentTerm + data[key];
+    return Object.keys(data).reduce(accumulator, '').toLowerCase();
+  }
+
+  constructor(initialData: T[] = []) {
+    this._data = new BehaviorSubject<T[]>(initialData);
+    this._updateChangeSubscription();
+  }
+
+  /**
+   * Subscribe to changes that should trigger an update to the table's rendered rows. When the
+   * changes occur, process the current state of the filter, sort, and pagination along with
+   * the provided base data and send it to the table for rendering.
+   */
+  _updateChangeSubscription() {
+    // Sorting and/or pagination should be watched if MdSort and/or MdPaginator are provided.
+    // Otherwise, use an empty observable stream to take their place.
+    const sortChange = this._sort ? this._sort.mdSortChange : observableOf();
+    const pageChange = this._paginator ? this._paginator.page : observableOf();
+
+    if (this._renderChangesSubscription) { this._renderChangesSubscription.unsubscribe(); }
+    this._renderChangesSubscription = RxChain.from(this._data)
+        // Watch for base data or filter changes to provide a filtered set of data.
+        .call(combineLatest, this._filter)
+        .call(map, ([data]) => this._filterData(data.slice()))
+        // Watch for filtered data or sort changes to provide an ordered set of data.
+        .call(combineLatest, startWith.call(sortChange, null))
+        .call(map, ([data]) => this._orderData(data.slice()))
+        // Watch for ordered data or page changes to provide a paged set of data.
+        .call(combineLatest, startWith.call(pageChange, null))
+        .call(map, ([data]) => this._pageData(data.slice()))
+        // Watched for paged data changes and send the result to the table to render.
+        .subscribe(data => this._renderData.next(data.slice()));
+  }
+
+  /**
+   * Returns a filtered data array where each filter object contains the filter string within
+   * the result of the filterTermAccessor function.
+   */
+  _filterData(data: T[]) {
+    // If there is a filter string, filter out data that does not contain it.
+    // Each data object is converted to a string using the function defined by filterTermAccessor.
+    // May be overriden for customization.
+    const filteredData = this.filter === '' ? data : data.filter(obj => {
```

Comment:
When filter is `null` / `undefined`? What about when filter is all whitespace?

Reply:
Changed to `!this.filter` to catch null/undefined. Not sure about whitespace, I think that our users should be doing these things according to their needs (e.g. `toLowerCase`, `trim`, etc.) before setting filter